saleor/graphql/discount/mutations/promotion/promotion_update.py

```diff
@@ -0,0 +1,140 @@
+        doc_category = DOC_CATEGORY_DISCOUNTS
+
+    @classmethod
+    def perform_mutation(cls, _root, info: ResolveInfo, /, **data):
+        instance = cls.get_instance(info, **data)
+        previous_end_date = instance.end_date
+        data: dict = data["input"]
+        manager = get_plugin_manager_promise(info.context).get()
+        cleaned_input: dict = cls.clean_input(info, instance, data)
+        with transaction.atomic():
+            instance = cls.construct_instance(instance, cleaned_input)
+            cls.clean_instance(info, instance)
+            cls.save(info, instance, cleaned_input)
+            cls._save_m2m(info, instance, cleaned_input)
+            # update the product undiscounted prices for promotion only when
+            # start or end date has changed
+            if "start_date" in cleaned_input or "end_date" in cleaned_input:
+                update_products_discounted_prices_of_promotion_task.delay(instance.pk)
+            cls.send_promotion_webhooks(
+                manager, instance, cleaned_input, previous_end_date
+            )
+        return cls.success_response(instance)
+
+    @classmethod
+    def clean_input(
+        cls, info: ResolveInfo, instance: models.Promotion, data: dict, **kwargs
+    ):
+        cleaned_input = super().clean_input(info, instance, data, **kwargs)
+        start_date = cleaned_input.get("start_date")
+        end_date = cleaned_input.get("end_date")
+        try:
+            validate_end_is_after_start(start_date, end_date)
+        except ValidationError as error:
+            error.code = PromotionUpdateErrorCode.INVALID.value
+            raise ValidationError({"endDate": error})
+        return cleaned_input
+
+    @classmethod
+    def send_promotion_webhooks(
+        cls,
+        manager: "PluginsManager",
+        instance: models.Promotion,
+        cleaned_input: dict,
+        previous_end_date: datetime,
+    ):
+        cls.call_event(
+            manager.promotion_updated,
+            instance,
+        )
+        cls.send_promotion_toggle_webhook(
+            manager, instance, cleaned_input, previous_end_date
+        )
+
+    @classmethod
+    def send_promotion_toggle_webhook(
+        cls,
+        manager: "PluginsManager",
+        instance: models.Promotion,
+        clean_input: dict,
+        previous_end_date: datetime,
+    ):
+        """Send a webhook about starting or ending promotion if it wasn't sent yet.
+
+        Send webhook when the start or end date already passed and the notification_date
+        is not set or the last notification was sent before start or end date.
+        """
+        now = datetime.now(pytz.utc)
+
+        notification_date = instance.last_notification_scheduled_at
+        start_date = clean_input.get("start_date")
+        end_date = clean_input.get("end_date")
+
+        if not start_date and not end_date:
+            return
+
+        send_notification = False
+        for date in [start_date, end_date]:
```

Comment:
I know that the PR is in draft but I saw it in different PR unrelated to this one. So I decided to leave a comment here.
What is the purpose of this `FOR`? I see we iterate over two dates: `start` and `end`, and set send_notification to True, but few lines below, we have a condition for setting `send_notification` for  `end_date`. Which means that end_date requires additional action. So maybe instead of `For` + additional `if` condition for `end_date`, we could have two `if`s - one for `start`, second one for `end`?

Reply:
We introduced `promotion_started` and `promotion_ended` webhook (instead of `promotion_toggle`), so I marked this comment as resolved.